apollo-router/src/plugins/authentication.rs

```diff
@@ -0,0 +1,944 @@
+                    //  - the retrieved JWKS is None (b)
+                    //  - we have a kid that we don't know about (c)
+                    // We need to do some additional processing.  (I've tagged with comments a/b/c below)
+                    let err_cleanup = move |cache_clear| {
+                        // If cache_clear is set, then we clear the cache so that subsequent
+                        // requests will retrieve new JWKS.
+                        // The COOLDOWN controls attempts to retrieve based on a new "kid", but not
+                        // repeated attempts due to failure to retrieve JWKS.
+                        if cache_clear {
+                            tracing::info!("Clearing cached JWKS");
+                            closure_jwks.clear();
+                            // Impose the COOLDOWN
+                            // Only spawn 1 task to remove the cooldown
+                            if COOLDOWN
+                                .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+                                .is_ok()
+                            {
+                                tokio::spawn(async move {
+                                    let t = my_config
+                                        .cooldown
+                                        .unwrap_or(DEFAULT_AUTHENTICATION_COOLDOWN);
+                                    tokio::time::sleep(t).await;
+                                    COOLDOWN.store(false, Ordering::SeqCst);
+                                });
+                            }
+                        }
+                    };
+
+                    let jwks_opt = match my_jwks.get(my_jwks_url).await {
+                        Ok(k) => k,
+                        Err(e) => {
+                            err_cleanup(false); // a.
+                            return failure_message(
+                                request.context,
+                                format!("Could not retrieve JWKS set: {e}"),
+                                StatusCode::INTERNAL_SERVER_ERROR, // XXX: Best error?
+                            );
+                        }
+                    };
+
+                    let jwks = match jwks_opt {
+                        Some(k) => k,
+                        None => {
+                            err_cleanup(false); // b.
+                            return failure_message(
+                                request.context,
+                                "Could not find JWKS set at the configured location".to_string(),
+                                StatusCode::INTERNAL_SERVER_ERROR,
+                            );
+                        }
+                    };
+
+                    // Now let's try to validate our token
+                    match jwks.find(&kid) {
+                        Some(jwk) => {
+                            let decoding_key = match DecodingKey::from_jwk(jwk) {
```

Comment:
could you extract the code to download the JWKS, deduplicate the call and handling the cooldown in one function like `find_jwks(jwks: Arc<Deduplicate<?>>, kid: &str) -> Result<DecodingKey, String>`?
It is a bit hard to follow here, with de JWKS downloading code, then JWT validation in the middle, then again JWKS download and cooldown code.

Reply:
I've re-factored the COOLDOWN code to move it to remove the closure and place the functionality in the only place where it's used now (unrecognised kid): 66935caf895ea2db237e9586dcb78866e6c956c8

So: I think this is much clearer and I don't think the proposed `find_jwks` is required. What do you think?

Reply:
this is way easier to follow, thanks